# Divvy Bike-Share: Cleaning/Manipulation of Data and Exploration

The data sets used here can be found [at this link](https://divvy-tripdata.s3.amazonaws.com/index.html).

In this Jupyter Notebook, we clean, manipulate, and explore Divvy Bike-Share data for the Google data analytics capstone project--more information can be found [here](https://kaylabollinger.github.io/port-pages/capstone.html).

First, we import the pandas library--this will be sufficient for our analysis.

In [1]:
import pandas as pd

## Load Data

To load the 12 separate data sets easily, we only need to rename the "202209" dataset from "202209-divvy-publictripdata" to match the convention "202209-divvy-tripdata" (to match the convention of the other datasets).

In [2]:
date_range = ['202111','202112','202201','202202','202203','202204','202205','202206','202207','202208','202209','202210']
df_list = []
for month in date_range:
    df_list.append(pd.read_csv(f'{month}-divvy-tripdata/{month}-divvy-tripdata.csv'))

Print out info on each dataset to confirm they contain the same attributes.

In [3]:
for df in df_list:
    print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359978 entries, 0 to 359977
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             359978 non-null  object 
 1   rideable_type       359978 non-null  object 
 2   started_at          359978 non-null  object 
 3   ended_at            359978 non-null  object 
 4   start_station_name  284688 non-null  object 
 5   start_station_id    284688 non-null  object 
 6   end_station_name    280791 non-null  object 
 7   end_station_id      280791 non-null  object 
 8   start_lat           359978 non-null  float64
 9   start_lng           359978 non-null  float64
 10  end_lat             359787 non-null  float64
 11  end_lng             359787 non-null  float64
 12  member_casual       359978 non-null  object 
dtypes: float64(4), object(9)
memory usage: 35.7+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 247540 entries, 0 to 247539
Dat

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785932 entries, 0 to 785931
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             785932 non-null  object 
 1   rideable_type       785932 non-null  object 
 2   started_at          785932 non-null  object 
 3   ended_at            785932 non-null  object 
 4   start_station_name  673895 non-null  object 
 5   start_station_id    673895 non-null  object 
 6   end_station_name    665410 non-null  object 
 7   end_station_id      665410 non-null  object 
 8   start_lat           785932 non-null  float64
 9   start_lng           785932 non-null  float64
 10  end_lat             785089 non-null  float64
 11  end_lng             785089 non-null  float64
 12  member_casual       785932 non-null  object 
dtypes: float64(4), object(9)
memory usage: 78.0+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 701339 entries, 0 to 701338
Dat

Since each dataset contains the same attributes, we can concatenate them into a single dataframe to analyze (their month information is still available in the "started_at"/"ended_at" attributes.

In [4]:
df = pd.concat(df_list)

Now we'll take a quick look at the data.

In [5]:
df

ride_id  rideable_type           started_at  \
0       7C00A93E10556E47  electric_bike  2021-11-27 13:27:38   
1       90854840DFD508BA  electric_bike  2021-11-27 13:38:25   
2       0A7D10CDD144061C  electric_bike  2021-11-26 22:03:34   
3       2F3BE33085BCFF02  electric_bike  2021-11-27 09:56:49   
4       D67B4781A19928D4  electric_bike  2021-11-26 19:09:28   
...                  ...            ...                  ...   
558680  BC3BFA659C9AB6F1   classic_bike  2022-10-30 01:41:29   
558681  ACD65450291CF95F   classic_bike  2022-10-30 01:41:54   
558682  4AAC03D1438E97CA   classic_bike  2022-10-15 09:34:11   
558683  8E6F3F29785E5D40   classic_bike  2022-10-09 10:21:34   
558684  8D14CBE672431920    docked_bike  2022-10-22 13:17:13   

                   ended_at          start_station_name start_station_id  \
0       2021-11-27 13:46:38                         NaN              NaN   
1       2021-11-27 13:56:10                         NaN              NaN   
2       2021-11-26 22:05:56                         NaN              NaN   
3       2021-11-27 10:01:50                         NaN              NaN   
4       2021-11-26 19:30:41                         NaN              NaN   
...                     ...                         ...              ...   
558680  2022-10-30 01:57:16  Clifton Ave & Armitage Ave     TA1307000163   
558681  2022-10-30 01:57:09  Clifton Ave & Armitage Ave     TA1307000163   
558682  2022-10-15 10:03:21     Sedgwick St & North Ave     TA1307000038   
558683  2022-10-09 10:43:45     Sedgwick St & North Ave     TA1307000038   
558684  2022-10-22 13:46:14     Clark St & Armitage Ave            13146   

                end_station_name end_station_id  start_lat  start_lng  \
0                            NaN            NaN  41.930000 -87.720000   
1                            NaN            NaN  41.960000 -87.700000   
2                            NaN            NaN  41.960000 -87.700000   
3                            NaN            NaN  41.940000 -87.790000   
4                            NaN            NaN  41.900000 -87.630000   
...                          ...            ...        ...        ...   
558680  Lincoln Ave & Roscoe St*   chargingstx5  41.918216 -87.656936   
558681  Lincoln Ave & Roscoe St*   chargingstx5  41.918216 -87.656936   
558682    Wabash Ave & Grand Ave   TA1307000117  41.911386 -87.638677   
558683  Damen Ave & Clybourn Ave          13271  41.911386 -87.638677   
558684    Wabash Ave & Grand Ave   TA1307000117  41.918306 -87.636282   

          end_lat    end_lng member_casual  
0       41.960000 -87.730000        casual  
1       41.920000 -87.700000        casual  
2       41.960000 -87.700000        casual  
3       41.930000 -87.790000        casual  
4       41.880000 -87.620000        casual  
...           ...        ...           ...  
558680  41.943350 -87.670668        casual  
558681  41.943350 -87.670668        casual  
558682  41.891466 -87.626761        casual  
558683  41.931931 -87.677856        member  
558684  41.891466 -87.626761        casual  

[5755694 rows x 13 columns]

We'll first convert "started_at" and "ended_at" to type "datetime", and then we will explore each attribute more closesly.

In [6]:
df['started_at'] = pd.to_datetime(df['started_at'])
df['ended_at'] = pd.to_datetime(df['ended_at'])

## ride_id

This attribute should be a unique identifier for each trip--we expect them to unique and not null, so we check for theses properties.

In [7]:
df[df['ride_id'].isna()]

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []

In [8]:
df[df['ride_id'].duplicated()]

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []

Everything looks good--nothing to fix/clean.

## rideable_type

This attribute should indicate what type of bike was used. We will first explore the different options described in the data.

In [9]:
df['rideable_type'].unique()

array(['electric_bike', 'classic_bike', 'docked_bike'], dtype=object)

There are no null values or apparent typos. However, it is not clear what the difference is between "electric_bike"/"classic_bike" vs "docked_bike". Normally this could be a quick question for a stakeholder, but for this case study we will just do our own investigation. First, we take a closer look at the "docked_bike" rideable_type.

In [10]:
df[df['rideable_type']=='docked_bike']

ride_id rideable_type          started_at  \
528     2C78CB8C153E2DFA   docked_bike 2021-11-07 14:39:41   
2032    860AFB2EA0ABC296   docked_bike 2021-11-06 14:55:37   
2768    4AEF5789ADF9D8ED   docked_bike 2021-11-04 04:09:57   
2777    15E0EAF943D03B1C   docked_bike 2021-11-13 13:36:24   
2788    655AF153783EFA54   docked_bike 2021-11-30 16:44:41   
...                  ...           ...                 ...   
558636  7C622DC6F652505F   docked_bike 2022-10-05 16:32:35   
558647  38D4A0162E2BE38D   docked_bike 2022-10-28 01:03:32   
558669  E632C8F3BB342576   docked_bike 2022-10-22 13:16:48   
558677  C5F29A096BDE2678   docked_bike 2022-10-22 13:17:09   
558684  8D14CBE672431920   docked_bike 2022-10-22 13:17:13   

                  ended_at         start_station_name start_station_id  \
528    2021-11-07 16:51:20      Michigan Ave & Oak St            13042   
2032   2021-11-06 16:24:24      Michigan Ave & Oak St            13042   
2768   2021-11-04 04:35:23       Paulina St & 18th St     TA1307000159   
2777   2021-11-13 13:43:56  Damen Ave & Charleston St            13288   
2788   2021-11-30 16:59:14             Burnham Harbor            15545   
...                    ...                        ...              ...   
558636 2022-10-05 17:15:36             Shedd Aquarium            15544   
558647 2022-10-28 10:06:36    Clark St & Armitage Ave            13146   
558669 2022-10-22 13:46:04    Clark St & Armitage Ave            13146   
558677 2022-10-22 13:46:18    Clark St & Armitage Ave            13146   
558684 2022-10-22 13:46:14    Clark St & Armitage Ave            13146   

                         end_station_name end_station_id  start_lat  \
528                 Michigan Ave & Oak St          13042  41.900960   
2032                Michigan Ave & Oak St          13042  41.900960   
2768              Morgan St & Pershing Rd          16933  41.857901   
2777              Logan Blvd & Elston Ave   TA1308000031  41.920082   
2788    DuSable Lake Shore Dr & Monroe St          13300  41.856268   
...                                   ...            ...        ...   
558636             Wabash Ave & Grand Ave   TA1307000117  41.867226   
558647             Wabash Ave & Grand Ave   TA1307000117  41.918306   
558669             Wabash Ave & Grand Ave   TA1307000117  41.918306   
558677             Wabash Ave & Grand Ave   TA1307000117  41.918306   
558684             Wabash Ave & Grand Ave   TA1307000117  41.918306   

        start_lng    end_lat    end_lng member_casual  
528    -87.623777  41.900960 -87.623777        casual  
2032   -87.623777  41.900960 -87.623777        casual  
2768   -87.668745  41.823613 -87.650931        casual  
2777   -87.677855  41.929465 -87.684158        casual  
2788   -87.613348  41.880958 -87.616743        casual  
...           ...        ...        ...           ...  
558636 -87.615355  41.891466 -87.626761        casual  
558647 -87.636282  41.891466 -87.626761        casual  
558669 -87.636282  41.891466 -87.626761        casual  
558677 -87.636282  41.891466 -87.626761        casual  
558684 -87.636282  41.891466 -87.626761        casual  

[182205 rows x 13 columns]

Docked bikes account for 3% of the total dataset, so the vast majority are labeled either "electric_bike" or "classic_bike". Since Divvy electric bikes are allowed to be parked anywhere--unlike the classic bikes which can only be parked at docking stations, it may be that "docked" bikes actually refer to classic bikes. This is also supported by the fact that up through July 2020, the only rideable_type listed in Divvy trip data was "docked_bike". Then [at the end of July 2020 electric bikes were introduced to the Divvy fleet](https://twitter.com/DivvyBikes/status/1288511684512870402), and the rideable_type "electric_bike" then appeared in the August 2020 data. This leads us to believe that "docked_bike" refers to "classic_bike", before the term "classic_bike" was adopted (as it was not needed before electric bikes were introduced). To support this hypothesis, we will take a look at the stations where "docked_bike"s could be found.

In [11]:
# check start stations for docked bikes
df[df['rideable_type']=='docked_bike']['start_station_name'].sort_values().unique()

array(['2112 W Peterson Ave', '63rd St Beach', '900 W Harrison St',
       'Aberdeen St & Jackson Blvd', 'Aberdeen St & Monroe St',
       'Aberdeen St & Randolph St', 'Ada St & 113th St',
       'Ada St & Washington Blvd', 'Adler Planetarium',
       'Albany Ave & 26th St', 'Albany Ave & Bloomingdale Ave',
       'Albany Ave & Montrose Ave', 'Altgeld Gardens',
       'Archer (Damen) Ave & 37th St', 'Artesian Ave & Hubbard St',
       'Ashland Ave & 13th St', 'Ashland Ave & 50th St',
       'Ashland Ave & 63rd St', 'Ashland Ave & 66th St',
       'Ashland Ave & 74th St', 'Ashland Ave & 78th St',
       'Ashland Ave & Archer Ave', 'Ashland Ave & Augusta Blvd',
       'Ashland Ave & Belle Plaine Ave', 'Ashland Ave & Blackhawk St',
       'Ashland Ave & Chicago Ave', 'Ashland Ave & Division St',
       'Ashland Ave & Garfield Blvd', 'Ashland Ave & Grace St',
       'Ashland Ave & Grand Ave', 'Ashland Ave & Lake St',
       'Ashland Ave & McDowell Ave', 'Ashland Ave & Pershing Rd',
       

In [12]:
# check start stations for classic bikes
df[df['rideable_type']=='classic_bike']['start_station_name'].sort_values().unique()

array(['2112 W Peterson Ave', '63rd St Beach', '900 W Harrison St',
       'Aberdeen St & Jackson Blvd', 'Aberdeen St & Monroe St',
       'Aberdeen St & Randolph St', 'Ada St & 113th St',
       'Ada St & Washington Blvd', 'Adler Planetarium',
       'Albany Ave & 26th St', 'Albany Ave & Bloomingdale Ave',
       'Albany Ave & Montrose Ave', 'Altgeld Gardens',
       'Archer (Damen) Ave & 37th St', 'Artesian Ave & Hubbard St',
       'Ashland Ave & 13th St', 'Ashland Ave & 50th St',
       'Ashland Ave & 63rd St', 'Ashland Ave & 66th St',
       'Ashland Ave & 74th St', 'Ashland Ave & 78th St',
       'Ashland Ave & Archer Ave', 'Ashland Ave & Augusta Blvd',
       'Ashland Ave & Belle Plaine Ave', 'Ashland Ave & Blackhawk St',
       'Ashland Ave & Chicago Ave', 'Ashland Ave & Division St',
       'Ashland Ave & Garfield Blvd', 'Ashland Ave & Grace St',
       'Ashland Ave & Grand Ave', 'Ashland Ave & Lake St',
       'Ashland Ave & McDowell Ave', 'Ashland Ave & Pershing Rd',
       

Looking through [Divvy's Twitter feed](https://twitter.com/DivvyBikes) after electric bikes were introduced, there are many posts about new electric bike stations being introduced, and it seems that none of them are listed in the stations where "classic" or "docked" bikes are picked up from. For example, [N Carpenter St & W Lake St](https://twitter.com/DivvyBikes/status/1338954379353673728) is listed as an E-station, and no classic/docked bikes were picked from here in the past year.

In [13]:
# check for electric bikes parked at N Carpenter St & W Lake St
df[(df['rideable_type']=='electric_bike') & (df['start_station_name']=='N Carpenter St & W Lake St')]

ride_id  rideable_type          started_at  \
2974    38C0FAF4EDDDCBB3  electric_bike 2021-11-23 17:12:03   
3905    8B3732C51A0E62FD  electric_bike 2021-11-06 17:34:57   
5659    F41E2AA0AD08E512  electric_bike 2021-11-08 12:52:28   
5688    3E4A13BAC8DE1675  electric_bike 2021-11-08 17:57:09   
9100    0D2496A4230375E7  electric_bike 2021-11-05 11:12:00   
...                  ...            ...                 ...   
535250  9F3389C11B7C3BB6  electric_bike 2022-10-24 18:23:41   
535901  294108D3F297F7B2  electric_bike 2022-10-12 14:38:32   
553942  D2725A2707D7B95E  electric_bike 2022-10-21 13:49:12   
553943  39E591BD67068B6E  electric_bike 2022-10-13 16:56:00   
553944  1E7097A0027C2859  electric_bike 2022-10-07 15:39:54   

                  ended_at          start_station_name start_station_id  \
2974   2021-11-23 17:16:04  N Carpenter St & W Lake St          20251.0   
3905   2021-11-06 17:47:30  N Carpenter St & W Lake St          20251.0   
5659   2021-11-08 12:57:44  N Carpenter St & W Lake St          20251.0   
5688   2021-11-08 18:02:32  N Carpenter St & W Lake St          20251.0   
9100   2021-11-05 11:12:38  N Carpenter St & W Lake St          20251.0   
...                    ...                         ...              ...   
535250 2022-10-24 18:25:19  N Carpenter St & W Lake St          20251.0   
535901 2022-10-12 14:52:50  N Carpenter St & W Lake St          20251.0   
553942 2022-10-21 13:52:46  N Carpenter St & W Lake St          20251.0   
553943 2022-10-13 17:11:16  N Carpenter St & W Lake St          20251.0   
553944 2022-10-07 15:43:34  N Carpenter St & W Lake St          20251.0   

                  end_station_name end_station_id  start_lat  start_lng  \
2974          Ogden Ave & Race Ave          13194      41.89     -87.65   
3905        State St & Randolph St   TA1305000029      41.89     -87.65   
5659                           NaN            NaN      41.89     -87.65   
5688                           NaN            NaN      41.89     -87.65   
9100           Morgan St & Lake St   TA1306000015      41.89     -87.65   
...                            ...            ...        ...        ...   
535250       Sangamon St & Lake St   TA1306000015      41.89     -87.65   
535901     Sedgwick St & North Ave   TA1307000038      41.89     -87.65   
553942  Aberdeen St & Jackson Blvd          13157      41.89     -87.65   
553943     Clark St & Armitage Ave          13146      41.89     -87.65   
553944  Aberdeen St & Jackson Blvd          13157      41.89     -87.65   

          end_lat    end_lng member_casual  
2974    41.891784 -87.658925        member  
3905    41.884461 -87.628197        casual  
5659    41.870000 -87.650000        member  
5688    41.880000 -87.650000        member  
9100    41.885512 -87.652492        member  
...           ...        ...           ...  
535250  41.885779 -87.651025        member  
535901  41.911386 -87.638677        member  
553942  41.877726 -87.654787        member  
553943  41.918306 -87.636282        member  
553944  41.877726 -87.654787        member  

[2554 rows x 13 columns]

In [14]:
# check for classic/dock bikes parked at N Carpenter St & W Lake St
df[~(df['rideable_type']=='electric_bike') & (df['start_station_name']=='N Carpenter St & W Lake St')]

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []

We will replace "docked_bike" with "classic_bike" in the dataset.

In [15]:
df.replace({'rideable_type':'docked_bike'},'classic_bike',inplace=True)

## Locations:
- start_station_name
- start_station_id
- end_station_name
- end_station_id
- start_lat
- start_lng
- end_lat
- end_lng

From just a quick glance of the dataset, we see that there are many null values in locational attributes. For some instances, it is ok to have a null value. Electric bikes can be parked anywhere, so we should expect that they are not always starting or ending at Divvy stations. Classic bikes on the other hand, should only be parked at Divvy stations--that is, they should not have null values for either their starting or ending position. We will take a closer look at where these attributes are null and decide what datapoints to keep.

A few things to check first before we dive in:
- if there is a lat value, is there also a lng value?
- if there is a name value, is there also an id value?

If the answers to both of these questions are yes, then we can narrow our search to just focusing on the name of the station and the lat of the station.

In [16]:
print(df[df['start_lat'].isna() & df['start_lng'].notnull()])
print(df[df['start_lat'].notnull() & df['start_lng'].isna()])
print(df[df['end_lat'].isna() & df['end_lng'].notnull()])
print(df[df['end_lat'].notnull() & df['end_lng'].isna()])

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []
Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []
Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []
Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []


In [17]:
print(df[df['start_station_name'].isna() & df['start_station_id'].notnull()])
print(df[df['start_station_name'].notnull() & df['start_station_id'].isna()])
print(df[df['end_station_name'].isna() & df['end_station_id'].notnull()])
print(df[df['end_station_name'].notnull() & df['end_station_id'].isna()])

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []
Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []
Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []
Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []


We can indeed focus our search on just station names and lat null values.

We will now check what types of bikes have null values for different positional attributes. First, we check null values for starting points:

In [18]:
# check names
df[df['start_station_name'].isna()]['rideable_type'].unique()

array(['electric_bike'], dtype=object)

In [19]:
# check lat
df[df['start_lat'].isna()]['rideable_type'].unique()

array([], dtype=object)

The null values for starting positions seem fine--only electric bikes have null values for their starting station name (which is allowed), but they always have lat/lng values (which we will need later in analysis).

Next we check null values for ending points.

In [20]:
# check names
df[df['end_station_name'].isna()]['rideable_type'].unique()

array(['electric_bike', 'classic_bike'], dtype=object)

In [21]:
# check lat
df[df['end_lat'].isna()]['rideable_type'].unique()

array(['classic_bike'], dtype=object)

We see that electric bikes may not have an end station name, but they always have an end station lat/lng--this is good and does not need fixing. We now explore the null values for classic bikes.

In [22]:
df[(df['end_station_name'].isna() | df['end_lat'].isna()) & (df['rideable_type']=='classic_bike')]

ride_id rideable_type          started_at  \
5450    B8D93992A39CE7B1  classic_bike 2021-11-06 10:04:44   
5523    AFF731A91A9C9440  classic_bike 2021-11-06 12:17:13   
5595    82F5BB7D74A20EBF  classic_bike 2021-11-10 16:01:10   
5670    D66FB7A5034CD519  classic_bike 2021-11-23 11:53:36   
5681    FBE4695DEE1A4BF9  classic_bike 2021-11-24 20:20:36   
...                  ...           ...                 ...   
547236  8C223882E19D0AB6  classic_bike 2022-10-09 10:14:13   
547306  E6B32143E49062FF  classic_bike 2022-10-01 04:59:23   
547309  3CA96CB9EC35FAA5  classic_bike 2022-10-25 00:46:50   
547695  6B912896DA81FE98  classic_bike 2022-10-09 18:37:00   
547777  71844616BB3E3BE3  classic_bike 2022-10-23 18:54:28   

                  ended_at               start_station_name start_station_id  \
5450   2021-11-06 10:16:19            Green St & Madison St     TA1307000120   
5523   2021-11-07 12:17:04          University Library (NU)              605   
5595   2021-11-11 08:43:40     Cornell Ave & Hyde Park Blvd     KA1503000007   
5670   2021-11-24 12:53:30         Laflin St & Cullerton St            13307   
5681   2021-11-25 07:39:31         Laflin St & Cullerton St            13307   
...                    ...                              ...              ...   
547236 2022-10-10 11:14:04          Clark St & Armitage Ave            13146   
547306 2022-10-02 05:59:18       Sheridan Rd & Montrose Ave     TA1307000107   
547309 2022-10-29 09:16:16             Rush St & Hubbard St     KA1503000044   
547695 2022-10-10 19:37:00  Monticello Ave & Irving Park Rd     KA1504000139   
547777 2022-10-24 19:54:16       Aberdeen St & Jackson Blvd            13157   

       end_station_name end_station_id  start_lat  start_lng  end_lat  \
5450                NaN            NaN  41.881892 -87.648789    41.88   
5523                NaN            NaN  42.052939 -87.673447    42.05   
5595                NaN            NaN  41.802406 -87.586924    41.81   
5670                NaN            NaN  41.854915 -87.663560      NaN   
5681                NaN            NaN  41.854915 -87.663560    41.87   
...                 ...            ...        ...        ...      ...   
547236              NaN            NaN  41.918306 -87.636282      NaN   
547306              NaN            NaN  41.961670 -87.654640      NaN   
547309              NaN            NaN  41.890173 -87.626185      NaN   
547695              NaN            NaN  41.954005 -87.719128      NaN   
547777              NaN            NaN  41.877726 -87.654787      NaN   

        end_lng member_casual  
5450     -87.63        member  
5523     -87.67        casual  
5595     -87.59        casual  
5670        NaN        casual  
5681     -87.66        member  
...         ...           ...  
547236      NaN        member  
547306      NaN        casual  
547309      NaN        casual  
547695      NaN        casual  
547777      NaN        member  

[6908 rows x 13 columns]

In [23]:
# group nulls by month to see if this is a problem with all datasets or just a few
df[(df['end_station_name'].isna() | df['end_lat'].isna()) & (df['rideable_type']=='classic_bike')].groupby(pd.Grouper(key='started_at',freq='M')).count()

ride_id  rideable_type  ended_at  start_station_name  \
started_at                                                         
2021-11-30      458            458       458                 458   
2021-12-31      413            413       413                 413   
2022-01-31      388            388       388                 388   
2022-02-28      208            208       208                 208   
2022-03-31      283            283       283                 283   
2022-04-30      324            324       324                 324   
2022-05-31      734            734       734                 734   
2022-06-30     1079           1079      1079                1079   
2022-07-31      953            953       953                 953   
2022-08-31      851            851       851                 851   
2022-09-30      731            731       731                 731   
2022-10-31      486            486       486                 486   

            start_station_id  end_station_name  end_station_id  start_lat  \
started_at                                                                  
2021-11-30               458                 0               0        458   
2021-12-31               413                 0               0        413   
2022-01-31               388                 0               0        388   
2022-02-28               208                 0               0        208   
2022-03-31               283                 0               0        283   
2022-04-30               324                 0               0        324   
2022-05-31               734                 0               0        734   
2022-06-30              1079                 0               0       1079   
2022-07-31               953                 0               0        953   
2022-08-31               851                 0               0        851   
2022-09-30               731                 0               0        731   
2022-10-31               486                 0               0        486   

            start_lng  end_lat  end_lng  member_casual  
started_at                                              
2021-11-30        458      267      267            458  
2021-12-31        413      269      269            413  
2022-01-31        388      302      302            388  
2022-02-28        208      131      131            208  
2022-03-31        283       17       17            283  
2022-04-30        324        7        7            324  
2022-05-31        734       12       12            734  
2022-06-30       1079       24       24           1079  
2022-07-31        953        6        6            953  
2022-08-31        851        8        8            851  
2022-09-30        731       19       19            731  
2022-10-31        486       11       11            486

There are 6908 instances of null values for end position for classic and docked bikes, and it does not seem to be due to a particular dataset (defined by month) missing these values. Since these instances only account for 0.1% of the total data, we should be able to remove them from the dataset and still be able to perform a meaningful analysis.

In [24]:
df = df[~((df['end_station_name'].isna() | df['end_lat'].isna()) & (df['rideable_type']=='classic_bike'))]
df.reset_index(drop=True,inplace=True)

Finally, we will explore the lat/lng values a bit more thoroughly. The main thing we want to check is that if a two different lat/lng values are associated with the same station, then those sets of values should be close to one another. Since the reported lat/lng values may not be very precise, we are mostly looking for anything that really stands out as different from the rest.

In [25]:
# make a new datafram containing only relevant start station information for this exploration
df_start_stations = df[(df['start_station_name'].notnull()) & (df['start_lat'].notnull())][['start_station_name','start_lat','start_lng']].copy()
df_start_stations.columns = ['name','lat','lng']

In [26]:
df_start_stations

name        lat        lng
5             Michigan Ave & Oak St  41.900864 -87.623794
45            Michigan Ave & Oak St  41.900960 -87.623777
49         Kingsbury St & Kinzie St  41.889131 -87.638437
76          Dearborn St & Monroe St  41.880793 -87.629458
95          Dearborn St & Monroe St  41.880788 -87.629916
...                             ...        ...        ...
5748781  Clifton Ave & Armitage Ave  41.918216 -87.656936
5748782  Clifton Ave & Armitage Ave  41.918216 -87.656936
5748783     Sedgwick St & North Ave  41.911386 -87.638677
5748784     Sedgwick St & North Ave  41.911386 -87.638677
5748785     Clark St & Armitage Ave  41.918306 -87.636282

[4870609 rows x 3 columns]

In [27]:
# make a new datafram containing only relevant end station information for this exploration
df_end_stations = df[df['end_station_name'].notnull() & df['end_lat'].notnull()][['end_station_name','end_lat','end_lng']].copy()
df_end_stations.columns = ['name','lat','lng']

In [28]:
# concatenate start and end station information
df_stations = pd.concat([df_start_stations,df_end_stations])
df_stations.drop_duplicates(inplace=True)

In [29]:
# check to see that there is indeed variance in the lat/lng values
df_stations[df_stations.duplicated(subset=['name'])].sort_values('name')

name        lat        lng
3942480        2112 W Peterson Ave  41.991097 -87.683570
4645887        2112 W Peterson Ave  41.991159 -87.683570
3806146        2112 W Peterson Ave  41.991177 -87.683588
2062822        2112 W Peterson Ave  41.991219 -87.683610
2077783        2112 W Peterson Ave  41.991223 -87.683585
...                            ...        ...        ...
4226650       Yates Blvd & 93rd St  41.726129 -87.566370
212925        Yates Blvd & 93rd St  41.726170 -87.566260
3744493       Yates Blvd & 93rd St  41.726157 -87.566404
1512699  Yates Blvd & Exchange Ave  41.760000 -87.570000
2719950             Zapata Academy  41.840000 -87.730000

[2106481 rows x 3 columns]

In [30]:
def get_loc(coord):
    d = {}
    d['lat_mean'] = coord['lat'].mean()
    d['lat_std'] = coord['lat'].std()
    d['lng_mean'] = coord['lng'].mean()
    d['lng_std'] = coord['lng'].std()
    return pd.Series(d)

In [31]:
df_gb = df_stations.groupby('name').apply(get_loc) # gb = groupby

In [32]:
df_gb.sort_values(['lat_std','lng_std'],ascending=[False,False])

lat_mean   lat_std  \
name                                                                    
Pawel Bialowas - Test- PBSC charging station      43.749817  2.666100   
DIVVY CASSETTE REPAIR MOBILE STATION              41.865848  0.076150   
WEST CHI-WATSON                                   41.883362  0.063952   
WestChi                                           41.894039  0.062494   
Base - 2132 W Hubbard Warehouse                   41.881889  0.062482   
...                                                     ...       ...   
Wolcott Ave & 61st St                             41.780000       NaN   
Wood St & 103rd St                                41.710000       NaN   
Wood St & 47th St                                 41.810000       NaN   
Woodlawn & 103rd - Olive Harvey Vaccination Site  41.710000       NaN   
Yale Ave & 119th St                               41.680000       NaN   

                                                   lng_mean   lng_std  
name                                                                   
Pawel Bialowas - Test- PBSC charging station     -80.738738  9.817840  
DIVVY CASSETTE REPAIR MOBILE STATION             -87.648313  0.047251  
WEST CHI-WATSON                                  -87.738881  0.044571  
WestChi                                          -87.728843  0.056490  
Base - 2132 W Hubbard Warehouse                  -87.706121  0.045875  
...                                                     ...       ...  
Wolcott Ave & 61st St                            -87.670000       NaN  
Wood St & 103rd St                               -87.670000       NaN  
Wood St & 47th St                                -87.670000       NaN  
Woodlawn & 103rd - Olive Harvey Vaccination Site -87.590000       NaN  
Yale Ave & 119th St                              -87.630000       NaN  

[1678 rows x 4 columns]

One station that clearly stands out from the others is "Pawel Bialowas - Test- PBSC charging station", so we will investigate this station.

In [33]:
df[(df['start_station_name']=='Pawel Bialowas - Test- PBSC charging station') | (df['end_station_name']=='Pawel Bialowas - Test- PBSC charging station')]

ride_id  rideable_type          started_at  \
618408  3327172413547F64  electric_bike 2022-01-14 11:13:15   

                  ended_at                            start_station_name  \
618408 2022-01-14 11:15:50  Pawel Bialowas - Test- PBSC charging station   

                                    start_station_id  \
618408  Pawel Bialowas - Test- PBSC charging station   

                                    end_station_name  \
618408  Pawel Bialowas - Test- PBSC charging station   

                                      end_station_id  start_lat  start_lng  \
618408  Pawel Bialowas - Test- PBSC charging station  45.635034 -73.796477   

        end_lat  end_lng member_casual  
618408  41.8646  -87.681        casual

This "test" station only accounts for a single data point, so we will just remove it.

In [34]:
df = df[~(df['start_station_name']=='Pawel Bialowas - Test- PBSC charging station')]
df.reset_index(drop=True,inplace=True)

## started_at / ended_at

First, we check for null values and convert to datetime.

In [35]:
df[df['started_at'].isna()]

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []

In [36]:
df[df['ended_at'].isna()]

Empty DataFrame
Columns: [ride_id, rideable_type, started_at, ended_at, start_station_name, start_station_id, end_station_name, end_station_id, start_lat, start_lng, end_lat, end_lng, member_casual]
Index: []

It would also be useful to consider the trip duration, so we will add this attribute to the dataframe.

In [37]:
df.loc[:,['trip_duration_sec']] = (df['ended_at'] - df['started_at']).dt.total_seconds()

As a sanity check, we check if there are any negative "trip_duration" values (as this would imply the "ended_at" time was before the "started_at" time--clearly an issue).

In [38]:
df[df['trip_duration_sec']<0]

ride_id  rideable_type          started_at  \
10177    B029250A1EFF2975   classic_bike 2021-11-07 01:40:02   
11033    D631251FA9C7FC03   classic_bike 2021-11-07 01:52:53   
12068    021DC77C70A3E367   classic_bike 2021-11-07 01:40:13   
23942    235ACD294AFB837F  electric_bike 2021-11-07 01:34:03   
29032    6A2DCA5CB1596CA6   classic_bike 2021-11-07 01:54:25   
...                   ...            ...                 ...   
5177233  57D8340BB2C0E7B0  electric_bike 2022-09-08 16:04:03   
5200870  F9A1F8F99C1EFBF5  electric_bike 2022-10-03 08:55:01   
5336732  5ACA942162006249  electric_bike 2022-10-21 19:29:00   
5344837  918F745F62CAC29E   classic_bike 2022-10-13 14:42:10   
5562907  D948CEECBD356EAB  electric_bike 2022-10-24 17:03:29   

                   ended_at           start_station_name start_station_id  \
10177   2021-11-07 01:05:46     Halsted St & Dickens Ave            13192   
11033   2021-11-07 01:05:22        Clark St & Newport St              632   
12068   2021-11-07 01:00:29         New St & Illinois St     TA1306000013   
23942   2021-11-07 01:17:13   Sheridan Rd & Lawrence Ave     TA1309000041   
29032   2021-11-07 01:03:44    Franklin St & Illinois St              RN-   
...                     ...                          ...              ...   
5177233 2022-09-08 16:01:09                          NaN              NaN   
5200870 2022-10-03 08:54:45    Chicago Ave & Sheridan Rd             E008   
5336732 2022-10-21 19:28:59                          NaN              NaN   
5344837 2022-10-13 11:53:28  Wilton Ave & Diversey Pkwy*     chargingstx0   
5562907 2022-10-24 17:03:28                          NaN              NaN   

                             end_station_name end_station_id  start_lat  \
10177                Leavitt St & Division St            658  41.919936   
11033              Racine Ave & Fullerton Ave   TA1306000026  41.944540   
12068                   Michigan Ave & 8th St            623  41.890847   
23942    Damen Ave & Thomas St (Augusta Blvd)   TA1307000070  41.969481   
29032     Mies van der Rohe Way & Chicago Ave          13338  41.891023   
...                                       ...            ...        ...   
5177233             Milwaukee Ave & Grand Ave          13033  41.890000   
5200870             Chicago Ave & Sheridan Rd           E008  42.050523   
5336732               Green St & Randolph St*   chargingstx3  41.880000   
5344837           Wilton Ave & Diversey Pkwy*   chargingstx0  41.932418   
5562907                                   NaN            NaN  41.940000   

         start_lng    end_lat    end_lng member_casual  trip_duration_sec  
10177   -87.648830  41.902997 -87.683825        casual            -2056.0  
11033   -87.654678  41.925563 -87.658404        member            -2851.0  
12068   -87.618617  41.872773 -87.623981        casual            -2384.0  
23942   -87.654727  41.901426 -87.677428        member            -1010.0  
29032   -87.635480  41.896910 -87.621743        casual            -3041.0  
...            ...        ...        ...           ...                ...  
5177233 -87.650000  41.891578 -87.648384        member             -174.0  
5200870 -87.677829  42.050491 -87.677821        member              -16.0  
5336732 -87.650000  41.883602 -87.648627        casual               -1.0  
5344837 -87.652705  41.932418 -87.652705        member           -10122.0  
5562907 -87.670000  41.940000 -87.670000        member               -1.0  

[112 rows x 14 columns]

Looks like there are some instances of having a negative trip duration, but there are so few of them that we can just remove them from the dataset.

In [39]:
df = df[df['trip_duration_sec']>=0]

Next, it would be meaningful to remove any trips that we will call "false starts"--that is, trips we will identify as "the user tried to use the bike but something went wrong, so they quickly put the bike back". We will identify these occurances as trips that have a short duration (less than 2 minutes) and have the same start/stop location (their start/end lat/lng values are within 0.001 of eachother). First, let's see how many of these "false starts" are in the data.

In [40]:
df[(df['trip_duration_sec']<120) & (abs(df['start_lat']-df['end_lat'])<0.001) & (abs(df['start_lng']-df['end_lng'])<0.001)]

ride_id  rideable_type          started_at  \
7        17069CC749126036  electric_bike 2021-11-27 14:33:56   
21       E7FA1E28D495677C  electric_bike 2021-11-02 18:57:21   
54       9724A6293110241C  electric_bike 2021-11-06 15:25:05   
66       281EAFBF5C4A3D2D  electric_bike 2021-11-06 11:48:42   
86       AEAC27FF919CEA85  electric_bike 2021-11-01 14:02:42   
...                   ...            ...                 ...   
5748307  BAFC9FE364030D41  electric_bike 2022-10-17 15:02:24   
5748367  DB76DE35ADA1CB07  electric_bike 2022-10-16 13:13:38   
5748368  4BFCD4F64A2BEF3B  electric_bike 2022-10-23 18:12:00   
5748374  27810887F8B837FA  electric_bike 2022-10-03 18:08:55   
5748375  A69055246F869E16  electric_bike 2022-10-17 06:48:21   

                   ended_at          start_station_name start_station_id  \
7       2021-11-27 14:34:38                         NaN              NaN   
21      2021-11-02 18:59:10                         NaN              NaN   
54      2021-11-06 15:26:12                         NaN              NaN   
66      2021-11-06 11:48:56                         NaN              NaN   
86      2021-11-01 14:03:58                         NaN              NaN   
...                     ...                         ...              ...   
5748307 2022-10-17 15:03:57        Rush St & Hubbard St     KA1503000044   
5748367 2022-10-16 13:15:01  Sheridan Rd & Montrose Ave     TA1307000107   
5748368 2022-10-23 18:12:37      Halsted St & Roscoe St     TA1309000025   
5748374 2022-10-03 18:10:21  Sheridan Rd & Montrose Ave     TA1307000107   
5748375 2022-10-17 06:48:44  Aberdeen St & Jackson Blvd            13157   

                   end_station_name end_station_id  start_lat  start_lng  \
7                               NaN            NaN  41.950000 -87.660000   
21                              NaN            NaN  41.890000 -87.770000   
54                              NaN            NaN  41.920000 -87.740000   
66                              NaN            NaN  41.800000 -87.690000   
86                              NaN            NaN  41.870000 -87.640000   
...                             ...            ...        ...        ...   
5748307        Rush St & Hubbard St   KA1503000044  41.890130 -87.626244   
5748367  Sheridan Rd & Montrose Ave   TA1307000107  41.961618 -87.654663   
5748368      Halsted St & Roscoe St   TA1309000025  41.943689 -87.648958   
5748374  Sheridan Rd & Montrose Ave   TA1307000107  41.961584 -87.654667   
5748375  Aberdeen St & Jackson Blvd          13157  41.877712 -87.654872   

           end_lat    end_lng member_casual  trip_duration_sec  
7        41.950000 -87.660000        casual               42.0  
21       41.890000 -87.770000        casual              109.0  
54       41.920000 -87.740000        casual               67.0  
66       41.800000 -87.690000        casual               14.0  
86       41.870000 -87.640000        casual               76.0  
...            ...        ...           ...                ...  
5748307  41.890173 -87.626185        member               93.0  
5748367  41.961670 -87.654640        member               83.0  
5748368  41.943670 -87.648950        member               37.0  
5748374  41.961670 -87.654640        member               86.0  
5748375  41.877726 -87.654787        member               23.0  

[141104 rows x 14 columns]

These "false starts" account for about 2.5% of the dataset. Since these "false starts" are not reporting meaningful use of the bikes, we will remove them from the dataset.

In [41]:
df = df[~((df['trip_duration_sec']<120) & (abs(df['start_lat']-df['end_lat'])<0.001) & (abs(df['start_lng']-df['end_lng'])<0.001))]
df.reset_index(drop=True,inplace=True)

## member_casual

We expect to see only "casual" and "member" as options here.

In [42]:
df['member_casual'].unique()

array(['casual', 'member'], dtype=object)

Looks good.

# Save Cleaned Data

In [43]:
df.to_csv('divvy_cleaned.csv')

# Explore Data Further--Look for Trends

To make things easy, we split the data into two sets--one for casual riders and one for member riders.

In [44]:
df_c = df[df['member_casual']=='casual']

In [45]:
df_m = df[df['member_casual']=='member'] 

## Bike Preference?
Check frequency.

In [46]:
# casual
df_c['rideable_type'].value_counts(normalize=True)

electric_bike    0.53978
classic_bike     0.46022
Name: rideable_type, dtype: float64

In [47]:
# member
df_m['rideable_type'].value_counts(normalize=True)

classic_bike     0.514821
electric_bike    0.485179
Name: rideable_type, dtype: float64

Casual riders have a slight preference for electric bikes (perhaps due to their more convenient pickup/dropoff allowed locations). Member riders may be more likely to have typical routes they need the bike for, and so may prefer a consistent pickup/dropoff location. Also, [only classic bikes are free for members](https://divvybikes.com/)--so members may be more inclined to want to take advantage of this.

## Preferred Day of Week?

Check frequency.

Note: 0 = Monday; 6 = Sunday

In [48]:
# casual
df_c['started_at'].dt.dayofweek.value_counts(normalize=True).sort_index()

0    0.121164
1    0.112479
2    0.117206
3    0.130579
4    0.144026
5    0.206022
6    0.168524
Name: started_at, dtype: float64

In [49]:
# member
df_m['started_at'].dt.dayofweek.value_counts(normalize=True).sort_index()

0    0.143957
1    0.154135
2    0.155808
3    0.156645
4    0.140094
5    0.133267
6    0.116094
Name: started_at, dtype: float64

Casual riders prefer weekend days, Saturday in particular (perhaps for fun weekend trips). Members prefer weekdays (perhaps for commuting).

## Preferred Month?

Check frequency.

In [50]:
# casual
df_c['started_at'].dt.month.value_counts(normalize=True).sort_index()

1     0.007856
2     0.009097
3     0.038358
4     0.053887
5     0.119339
6     0.156881
7     0.172305
8     0.152233
9     0.125899
10    0.088487
11    0.045792
12    0.029864
Name: started_at, dtype: float64

In [51]:
# member
df_m['started_at'].dt.month.value_counts(normalize=True).sort_index()

1     0.025088
2     0.027601
3     0.057188
4     0.071911
5     0.104229
6     0.117840
7     0.122604
8     0.125421
9     0.118819
10    0.102585
11    0.074379
12    0.052337
Name: started_at, dtype: float64

Everyone prefers warm months, but members are slightly more likely to use during colder months.

## Typical Trip Duration?

In [52]:
# casual
df_c['trip_duration_sec'].describe()

count    2.287917e+06
mean     1.370238e+03
std      4.624728e+03
min      0.000000e+00
25%      4.690000e+02
50%      8.100000e+02
75%      1.480000e+03
max      2.061244e+06
Name: trip_duration_sec, dtype: float64

In [53]:
# member
df_m['trip_duration_sec'].describe()

count    3.319652e+06
mean     7.585186e+02
std      1.094299e+03
min      0.000000e+00
25%      3.200000e+02
50%      5.430000e+02
75%      9.320000e+02
max      8.957500e+04
Name: trip_duration_sec, dtype: float64

Members have a smaller median trip duration--perhaps because they are more likely to have a specific desitination in mind (rather than wandering). Members also have smaller variance in trip duration.

Check trip duration on different days.

In [54]:
# casual
df_c.groupby(df_c['started_at'].dt.dayofweek).median()['trip_duration_sec']

started_at
0    803.0
1    720.0
2    714.0
3    727.0
4    774.0
5    929.0
6    935.0
Name: trip_duration_sec, dtype: float64

In [55]:
# member
df_m.groupby(df_m['started_at'].dt.dayofweek).median()['trip_duration_sec']

started_at
0    520.0
1    519.0
2    527.0
3    530.0
4    535.0
5    608.0
6    590.0
Name: trip_duration_sec, dtype: float64

Both casual riders and members tend to use the bike longer on weekends.

## Preferred Stations?

### Start Stations

In [56]:
# casual
df_c['start_station_name'].value_counts(normalize=True).sort_values(ascending=False)

Streeter Dr & Grand Ave                       2.914563e-02
DuSable Lake Shore Dr & Monroe St             1.628174e-02
Millennium Park                               1.295596e-02
Michigan Ave & Oak St                         1.271833e-02
DuSable Lake Shore Dr & North Blvd            1.192314e-02
                                                  ...     
Public Rack - Austin Ave & Wellington Ave     5.143498e-07
Public Rack - Kedzie Ave & 62nd Pl            5.143498e-07
City Rack - Albany Ave & 103rd St             5.143498e-07
Public Rack - Kedzie Ave &amp; 62nd Pl        5.143498e-07
Public Rack - Rutherford Ave & Belmont Ave    5.143498e-07
Name: start_station_name, Length: 1560, dtype: float64

In [57]:
# member
df_m['start_station_name'].value_counts(normalize=True).sort_values(ascending=False)

Kingsbury St & Kinzie St              8.955828e-03
Clark St & Elm St                     7.857651e-03
Wells St & Concord Ln                 7.606973e-03
University Ave & 57th St              7.027478e-03
Ellis Ave & 60th St                   6.918934e-03
                                          ...     
Public Rack - Harper Ave & 63rd St    3.535660e-07
Sawyer Ave & Bryn Mawr Ave - SE       3.535660e-07
Michigan Ave & 102nd St               3.535660e-07
Green St & 103rd St                   3.535660e-07
Public Rack - Kedzie Ave & 58th St    3.535660e-07
Name: start_station_name, Length: 1459, dtype: float64

### end stations

In [58]:
df_c['end_station_name'].value_counts(normalize=True).sort_values(ascending=False)

Streeter Dr & Grand Ave                       3.117026e-02
DuSable Lake Shore Dr & Monroe St             1.547482e-02
Millennium Park                               1.405679e-02
Michigan Ave & Oak St                         1.380914e-02
DuSable Lake Shore Dr & North Blvd            1.363626e-02
                                                  ...     
Public Rack - Bishop St & 81st St             5.303048e-07
Public Rack - Keef Ave & South Chicago Ave    5.303048e-07
Public Rack - Calumet Ave & Garfield Blvd     5.303048e-07
Public Rack - Brainerd Park                   5.303048e-07
Public Rack - Halsted St & 63rd St            5.303048e-07
Name: end_station_name, Length: 1575, dtype: float64

In [59]:
df_m['end_station_name'].value_counts(normalize=True).sort_values(ascending=False)

Kingsbury St & Kinzie St                  8.763193e-03
Clark St & Elm St                         7.987852e-03
Wells St & Concord Ln                     7.809198e-03
University Ave & 57th St                  7.292657e-03
Clinton St & Washington Blvd              7.156019e-03
                                              ...     
Public Rack - Menard Ave & Adams Blvd     3.530698e-07
Public Rack - Kedzie Ave & Pershing Rd    3.530698e-07
Public Rack - Kedzie Ave & 61st Pl W      3.530698e-07
Public Rack - Kedzie & 103rd St - W       3.530698e-07
Public Rack - Kedzie Ave & 58th St        3.530698e-07
Name: end_station_name, Length: 1434, dtype: float64

Casual riders tend to have destinations related to leisure, members tend to have desitions related to work.